File: core/src/main/java/com/uiptv/util/AppLog.java
package com.uiptv.util;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class AppLog {
    private static final List<Consumer<String>> listeners = new CopyOnWriteArrayList<>();
    private static final int MAX_LOG_LENGTH = 4000;
    private static final String SHOW_LOGS_PROPERTY = "uiptv.showLogs";
    private static volatile boolean terminalLoggingEnabled =
            Boolean.parseBoolean(System.getProperty(SHOW_LOGS_PROPERTY, "false"));

    private AppLog() {
    }

    public static void addInfoLog(Class<?> logSource, String log) {
        logWithLevel(logSource, log, LogLevel.INFO);
    }

    public static void addWarningLog(Class<?> logSource, String log) {
        logWithLevel(logSource, log, LogLevel.WARNING);
    }

    public static void addErrorLog(Class<?> logSource, String log) {
        logWithLevel(logSource, log, LogLevel.ERROR);
    }

    public static void addErrorLog(Class<?> logSource, String log, Throwable throwable) {
        logWithLevel(logSource, log, LogLevel.ERROR, throwable);
    }

    public static void addLog(Class<?> logSource, String log) {
        addInfoLog(logSource, log);
    }

    private static void logWithLevel(Class<?> logSource, String log, LogLevel level) {
        logWithLevel(logSource, log, level, null);
    }

    private static void logWithLevel(Class<?> logSource, String log, LogLevel level, Throwable throwable) {
        if (logSource == null) {
            throw new IllegalArgumentException("logSource cannot be null");
        }
        String safeLog = sanitizeLogMessage(log);
        if (isTerminalLoggingEnabled()) {
            Logger logger = LoggerFactory.getLogger(logSource);
            switch (level) {
                case ERROR -> {
                    if (throwable == null) {
                        logger.error(safeLog);
                    } else {
                        logger.error(safeLog, throwable);
                    }
                }
                case WARNING -> logger.warn(safeLog);
                case INFO -> logger.info(safeLog);
            }
        }
        for (Consumer<String> listener : listeners) {
            try {
                listener.accept(safeLog);
            } catch (Exception e) {
                // Keep logging resilient if a listener fails.
                Logger logger = LoggerFactory.getLogger(AppLog.class);
                logger.warn("Log listener failed: {}", e.toString());
            }
        }
    }

    public static String sanitizeValue(String value) {
        return sanitizeLogMessage(value);
    }

    private static String sanitizeLogMessage(String message) {
        if (message == null) {
            return "";
        }
        StringBuilder normalizedBuilder = new StringBuilder(message.length());
        for (int i = 0; i < message.length(); i++) {
            char current = message.charAt(i);
            normalizedBuilder.append(Character.isISOControl(current) ? ' ' : current);
        }
        String normalized = normalizedBuilder.toString().trim();
        if (normalized.length() <= MAX_LOG_LENGTH) {
            return normalized;
        }
        return normalized.substring(0, MAX_LOG_LENGTH) + "...";
    }

    public static void registerListener(Consumer<String> listener) {
        if (listener != null) {
            listeners.add(listener);
        }
    }

    public static void unregisterListener(Consumer<String> listener) {
        if (listener != null) {
            listeners.remove(listener);
        }
    }

    public static void setTerminalLoggingEnabled(boolean enabled) {
        terminalLoggingEnabled = enabled;
        System.setProperty(SHOW_LOGS_PROPERTY, Boolean.toString(enabled));
    }

    public static boolean isTerminalLoggingEnabled() {
        return terminalLoggingEnabled || Boolean.parseBoolean(System.getProperty(SHOW_LOGS_PROPERTY, "false"));
    }

    private enum LogLevel {
        INFO,
        WARNING,
        ERROR
    }
}

File: server/src/main/kotlin/com/uiptv/util/AppLog.kt
package com.uiptv.util

import org.slf4j.LoggerFactory
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer

object AppLog {
    private val listeners = CopyOnWriteArrayList<Consumer<String>>()
    private const val MAX_LOG_LENGTH = 4000
    private const val SHOW_LOGS_PROPERTY = "uiptv.showLogs"

    @JvmStatic
    fun addInfoLog(logSource: Class<*>, log: String?) {
        logWithLevel(logSource, log, LogLevel.INFO)
    }

    @JvmStatic
    fun addWarningLog(logSource: Class<*>, log: String?) {
        logWithLevel(logSource, log, LogLevel.WARNING)
    }

    @JvmStatic
    fun addErrorLog(logSource: Class<*>, log: String?) {
        logWithLevel(logSource, log, LogLevel.ERROR)
    }

    @JvmStatic
    fun addLog(logSource: Class<*>, log: String?) {
        addInfoLog(logSource, log)
    }

    private fun logWithLevel(logSource: Class<*>, log: String?, level: LogLevel) {
        requireNotNull(logSource) { "logSource cannot be null" }
        val safeLog = sanitizeLogMessage(log)
        if (isTerminalLoggingEnabled()) {
            val logger = LoggerFactory.getLogger(logSource)
            when (level) {
                LogLevel.ERROR -> logger.error(safeLog)
                LogLevel.WARNING -> logger.warn(safeLog)
                LogLevel.INFO -> logger.info(safeLog)
            }
        }
        for (listener in listeners) {
            try {
                listener.accept(safeLog)
            } catch (e: Exception) {
                LoggerFactory.getLogger(AppLog::class.java).error("Log listener failed", e)
            }
        }
    }

    @JvmStatic
    fun sanitizeValue(value: String?): String = sanitizeLogMessage(value)

    private fun sanitizeLogMessage(message: String?): String {
        if (message == null) {
            return ""
        }
        val normalized = buildString(message.length) {
            message.forEach { current -> append(if (Character.isISOControl(current)) ' ' else current) }
        }.trim()
        return if (normalized.length <= MAX_LOG_LENGTH) normalized else normalized.substring(0, MAX_LOG_LENGTH) + "..."
    }

    @JvmStatic
    fun registerListener(listener: Consumer<String>?) {
        if (listener != null) {
            listeners.add(listener)
        }
    }

    @JvmStatic
    fun unregisterListener(listener: Consumer<String>?) {
        if (listener != null) {
            listeners.remove(listener)
        }
    }

    private fun isTerminalLoggingEnabled(): Boolean =
        java.lang.Boolean.parseBoolean(System.getProperty(SHOW_LOGS_PROPERTY, "false"))

    private enum class LogLevel {
        INFO,
        WARNING,
        ERROR
    }
}
